### PR TITLE
chore: change undocumented protocol.registerProtocol to detect body type

### PR DESCRIPTION
**Pull request by nornagon**
#### Description of Change

This changes the API of the already existing but undocumented
protocol.registerProtocol function, which allows returning different
body data types. Marked semver/patch because this API is undocumented.

Recommend [reviewing with ?w=1](https://github.com/electron/electron/pull/36595/files?w=1).

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [ ] relevant documentation, tutorials, templates and examples are changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: none

### CODE REVIEW

**Comment by codebytere** [comment]
@nornagon for clarification - what's the primary motivation for this refactor? general simplification? I see that you added a future deprecation comment for only supporting `kFree` - is the impetus for that something upstream?

**Comment by nornagon** [comment]
@codebytere I have a broader goal of significantly refactoring the `protocol` module, roughly along these lines: https://github.com/electron/governance/pull/368

File: shell/browser/net/electron_url_loader_factory.h [hunk 1]
```diff
@@ -135,33 +135,27 @@ class ElectronURLLoaderFactory : public network::SelfDeletingURLLoaderFactory {
...
-      network::ResourceRequest request,
       mojo::PendingRemote<network::mojom::URLLoaderClient> client,
+      mojo::PendingReceiver<network::mojom::URLLoader> loader,
       network::mojom::URLResponseHeadPtr head,
-      const gin_helper::Dictionary& dict,
-      v8::Isolate* isolate,
-      v8::Local<v8::Value> response);
+      const network::ResourceRequest& original_request,
+      base::FilePath path,
+      gin_helper::Dictionary opts);
```
Comment (zcbenz) [inline shell/browser/net/electron_url_loader_factory.h:149]:
```suggestion
      const base::FilePath& path,
      const gin_helper::Dictionary& opts);
```

**Review by zcbenz — APPROVED** [review]

**Comment by zcbenz** [comment]
The function body needs to be changed too.

**Pull request merged**